spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/ReplicaPartitioner.scala

```diff
@@ -38,14 +62,20 @@ class ReplicaPartitioner(partitionsPerReplicaSet: Int, val connector: CassandraC
    */
   override def getPartition(key: Any): Int = {
     key match {
-      case key: Set[_] if key.size > 0 && key.forall(_.isInstanceOf[InetAddress]) =>
+      case key: T =>
         //Only use ReplicaEndpoints in the connected DC
-        val replicaSetInDC = (hosts.toSet & key.asInstanceOf[Set[InetAddress]]).toVector
+        val token = tokenGenerator.getTokenFor(key)
+        val tokenHash = Math.abs(token.hashCode())
+        val replicas = metadata
+          .getReplicas(keyspace, token.serialize(protocolVersion))
+          .map(_.getBroadcastAddress)
+
+        val replicaSetInDC = (hostSet & replicas).toVector
         if (replicaSetInDC.nonEmpty) {
-          val endpoint = replicaSetInDC(rand.nextInt(replicaSetInDC.size))
-          hostMap(endpoint)(rand.nextInt(partitionsPerReplicaSet))
+          val endpoint = replicaSetInDC(tokenHash % replicaSetInDC.size)
+          hostMap(endpoint)(tokenHash % partitionsPerReplicaSet)
         } else {
-          hostMap(randomHost)(rand.nextInt(partitionsPerReplicaSet))
+          hostMap(randomHost(tokenHash))(tokenHash % partitionsPerReplicaSet)
         }
       case _ => throw new IllegalArgumentException(
```

Comment:
Does it make any sense to keep it if we're matching all objects with the previous case?